### PR TITLE
feat(46907): Add `Intl.ListFormat` type declarations

### LINKED ISSUE #46907: Add Intl.ListFormat to es2021 lib
# Suggestion

## 🔍 Search Terms

- `Intl.ListFormat`
- `Intl`
- `ListFormat`

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Add `Intl.ListFormat` to es2021 lib types (as per [Finished Proposals](https://github.com/tc39/proposals/blob/master/ecma402/finished-proposals.md)).

Bringing this API to Typescript has been directly suggested before, but didn't follow the issue templates (#46866), and was also mentioned before (https://github.com/microsoft/TypeScript/issues/45646#issuecomment-962430370, https://github.com/microsoft/TypeScript/issues/29129#issuecomment-449608280), but, as I couldn't find any specific active issue where the feature suggestion is being considered, I'm opening this issue.

## 📃 Motivating Example

Code that uses the Intl.ListFormat API should not fail type checking:

```ts
new Intl.ListFormat('bg', {type: 'conjunction'}).format(['a','b','c'])
new Intl.ListFormat('bg', {type: 'disjunction'}).formatToParts(['a','b','c'])
```

## 💻 Use Cases

```ts
new Intl.ListFormat('bg', {type: 'conjunction'}).format(['a','b','c'])
new Intl.ListFormat('bg', {type: 'disjunction'}).formatToParts(['a','b','c'])
```

### PR DESCRIPTION
**Pull request by l2ig**
Fixes #46907

### CODE REVIEW

File: src/lib/es2021.intl.d.ts [hunk 1]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
...
+        format(list: Iterable<string>): string;
+
+        /**
+         * Returns an Array of objects representing the different components that can be used to format a list of values in a locale-aware fashion.
+         *
+         * @param list - An Array of values to be formatted according to a locale.
+         *
+         * @throws `TypeError` if `list` includes something other than the possible values.
+         *
+         * @returns {string} An Array of components which contains the formatted parts from the list.
```
Comment (l2ig) [inline src/lib/es2021.intl.d.ts:81]: Fix return type tomorrow

File: src/lib/es2021.intl.d.ts [hunk 2]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
...
+         *
+         * @param list - An Array of values to be formatted according to a locale.
+         *
+         * @throws `TypeError` if `list` includes something other than the possible values.
+         *
+         * @returns {{ type: "element" | "literal", value: string; }[]} An Array of components which contains the formatted parts from the list.
+         *
+         * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/formatToParts).
+         */
+        formatToParts(list: string[]): { type: "element" | "literal", value: string; }[];
```
Comment (andreubotella) [inline src/lib/es2021.intl.d.ts:85]: `formatToParts` takes an iterable of strings, not an array of strings. The MDN documentation here is wrong.

Comment (l2ig) [reply src/lib/es2021.intl.d.ts]: Good call :+1:, opened https://github.com/mdn/content/issues/11818

File: src/lib/es2021.intl.d.ts [hunk 3]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
...
+     *
+     * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/ListFormat#parameters).
+     */
+    interface ListFormatOptions {
+        /** The locale matching algorithm to use. For information about this option, see [Intl page](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_negotiation). */
+        localeMatcher?: ListFormatLocaleMatcher;
+        /** The format of output message. */
+        type?: ListFormatType;
+        /** The length of the internationalized message. */
+        style?: ListFormatStyle;
```
Comment (jakebailey) [inline src/lib/es2021.intl.d.ts:57]: These probably need `| undefined` stuck at the end; see the other object above for consistency.

File: src/lib/es2021.intl.d.ts [hunk 4]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
         formatRange(startDate: number | bigint, endDate: number | bigint): string;
         formatRangeToParts(startDate: number | bigint, endDate: number | bigint): NumberFormatPart[];
     }
-}
\ No newline at end of file
+
+    /**
+     * The locale matching algorithm to use.
+     *
+     * [MDN](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_negotiation).
```
Comment (jakebailey) [inline src/lib/es2021.intl.d.ts:28]: Not entirely certain it matters, but is the link used for the other parameter types below more useful? I suppose this one has more info about them, but the parameters page does link back to this one too.

File: src/lib/es2021.intl.d.ts [hunk 5]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
...
+         *
+         * @param list - An iterable object, such as an [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), to be formatted according to a locale.
+         *
+         * @throws `TypeError` if `list` includes something other than the possible values.
+         *
+         * @returns {{ type: "element" | "literal", value: string; }[]} An Array of components which contains the formatted parts from the list.
+         *
+         * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/formatToParts).
+         */
+        formatToParts(list: Iterable<string>): { type: "element" | "literal", value: string; }[];
```
Comment (jakebailey) [inline src/lib/es2021.intl.d.ts:85]: Should this return type be named somewhere?

**Review by jakebailey — CHANGES_REQUESTED** [review]
Seems correct, though you probably need the aforementioned `| undefined`.

File: src/lib/es2021.intl.d.ts (hunk 3 above)
Comment (l2ig) [reply src/lib/es2021.intl.d.ts:57]: But why if we already have `?`?

File: src/lib/es2021.intl.d.ts (hunk 5 above)
Comment (l2ig) [reply src/lib/es2021.intl.d.ts:85]: What do you prefer?

File: src/lib/es2021.intl.d.ts (hunk 3 above)
Comment (jakebailey) [reply src/lib/es2021.intl.d.ts:57]: For users of `exactOptionalPropertyTypes`.

- https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes
- https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#exact-optional-property-types
- #45652 and #45647 as examples.

File: src/lib/es2021.intl.d.ts (hunk 5 above)
Comment (jakebailey) [reply src/lib/es2021.intl.d.ts:85]: It's probably "safe" to leave it without a name and add one later. I can't think of a good name; the function refers to the return as "parts" but the doc refers to the return as "components", so...

File: src/lib/es2021.intl.d.ts (hunk 3 above)
Comment (jakebailey) [reply src/lib/es2021.intl.d.ts]: I will say that confusingly, `ResolvedDateTimeFormatOptions` is missing the `| undefined`. Not sure if that was intended or not in a previous PR or what.

File: src/lib/es2021.intl.d.ts [hunk 6]
```diff
@@ -21,4 +21,106 @@ declare namespace Intl {
         dayPeriod?: "narrow" | "short" | "long";
         fractionalSecondDigits?: 0 | 1 | 2 | 3;
     }
+
+    /**
+     * The locale matching algorithm to use.
+     *
+     * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/ListFormat#parameters).
+     */
+    type ListFormatLocaleMatcher = "lookup" | "best fit";
```
Comment (jakebailey) [inline src/lib/es2021.intl.d.ts:30]: One final comment: I'm noticing that none of the options above have defined types (or make use of the new ones); is that intentional? Is it more "correct" to not have defined types for any of them? Or do they need to be fixed to have defined types too?

(I'm also noticing that there are some silly things above like `"best fit" | "best fit"`.)

Comment (l2ig) [reply src/lib/es2021.intl.d.ts:30]: I'd rather not touch it in this PR, but create a follow-up PR in the next days instead.

* Create named types for the options above, remove duplicate `"best fit"`
* Investigate if `ResolvedDateTimeFormatOptions` is even correct (at a first glance, it seems wrong, but might be to match some unique browser behavior?)

Feel free to disagree, I can also tackle them here

Comment (jakebailey) [reply src/lib/es2021.intl.d.ts:30]: I'm sure a second PR is fine, I just wasn't sure (I have no major knowledge of these types) whether or not it's normal to have top level declared type names or not.

**Review by jakebailey — APPROVED** [review]

**Pull request merged**